Cleaned up the project and added the gradle support -- java 11

Mahmoud-Khawaja opened the pull request.

Mahmoud-Khawaja committed:
Cleaned up the project and added the gradle support -- java 11

Mahmoud-Khawaja committed:
Remove files and folders now ignored by .gitignore

Mahmoud-Khawaja commented:
@cyrille-artho 
Hello Cyrille, 
i deleted `src/classes/java/net` and `src/classes/java/io` cause they already exist in java standard lib. everything works fine for the gradle support now but we have two failing tests 

`testTimedoutAccept()` and `testDeadlockForBlockedFinalizer_MulitProcessVM()` 
im not sure what they fail. what do you think about it?

cyrille-artho commented:
Hi Mahmoud,
Thank you for starting to work on this.
As it turns out, we need a couple of files that you have marked for deletion.
* `bin/`: These are scripts needed to run/test JPF. They are not compiled from source, so we need to keep them.
* `src/classes/**/*.java`: These are the same classes as in the Java library, but their functionality is altered so they replace the original library classes when JPF is used. These classes are called "model classes" and very important for JPF being able to function correctly with I/O.

It seems that the unit tests for jpf-nas do not cover such behavior, which is something to look into. (It is a bit challenging to set up and run such tests, as these would have to be system tests.)

Check this presentation for the general architecture of JPF: https://docs.google.com/presentation/d/1K9DO8NuSZRSoQ8jLXClJJYBEJ-JGKD_8j27e11HWUUk/edit?usp=sharing

I can explain this more next week.

cyrille-artho reviewed (changes requested):
Hi Mahmoud,
Please check my comment; the files in `bin/` and `src/classes/` are essential, hence we have to keep them.
/ Cyrille

Mahmoud-Khawaja committed:
Added src/classes and bin/

Mahmoud-Khawaja commented:
Hi Cyrille,
Thank you so much for the clarification. Sorry for the late reply. I’ve added the files and look forward to meeting you in the next meeting :)

cyrille-artho reviewed (changes requested):
Looks good!
You have accidentally included `.gradle` and its contents in the PR; please remove that, as these files are downloaded by `./gradlew` and/or are lock files that have to be set/changed by Gradle.
Maybe the pattern `.gradle` in `.gitignore` does not match files inside `.gradle`.

Mahmoud-Khawaja committed:
Delete .gradle directory

Mahmoud-Khawaja committed:
fixed the gitignore

Mahmoud-Khawaja commented:
That was a typo in the .gitignore now it should be ignoring files correctly

cyrille-artho commented:
Thanks. The change looks good but I have a problem building jpf-core on this computer now, so I cannot try it out. It would be good to have a CI test workflow for GitHub in order to run some tests on the server.

cyrille-artho commented:
Can you please share your build log? On another computer, where jpf-core works, I get a lot of build errors:
```
: error: package exists in another module: java.base
package java.net;
```
and similar. I think the `--patch-module` option is needed, similar to how it is used in `jpf-core`.

Mahmoud-Khawaja committed:
fixed the build.gradle

Mahmoud-Khawaja commented:
The error was due to a conflict with the module system. when we tried to compile something like `classes/java/net/Socket.java`, it conflicted with the `java.base` module where that class already exists. i packaged the model classes into a separate jar file so they're loaded only by JPF  not the host jvm. now it works fine. 
![image](https://github.com/user-attachments/assets/0b2392d1-ea56-4104-9b6f-17da37d90e47)

cyrille-artho commented on build.gradle:
```diff
@@ -0,0 +1,104 @@
+
+dependencies {
+    // JPF-core jars (adjust path as needed)
+    implementation files("/home/finn/jpf-push/jpf-core/build/jpf.jar")
```
Please make sure that paths are relative, as automated builds cannot "adjust the path as needed".

cyrille-artho commented on jpf.properties:
```diff
@@ -1,40 +1,10 @@
-# since both do not support wildcard patterns in classpath specs
 jpf-nas.native_classpath =\
-   ${jpf-nas}/build/jpf-nas.jar;\
+   ${jpf-nas}/build/libs/jpf-nas.jar;\
```
The old build produced two JAR file with classes; one at the JPF-native level (JPF's "Java Model Interface"), and one at the library level (model classes).
Please try to fix the build, and once you have the path names resolved, I can try to run the new version on an example to see if this setup works.

cyrille-artho reviewed (changes requested):
Hi,
Please fix the absolute path names and consider reinstating two sets of library classes, so the right classes are loaded for the right purpose. This is a bit complicated; you can find more details on the wiki or in this publication: https://link.springer.com/chapter/10.1007/978-3-031-57249-4_1
/ Cyrille

Mahmoud-Khawaja committed:
fix build.gradle and added the relative path

Mahmoud-Khawaja commented:
i fixed the absolute path issue and modified the build to compilee the model classes using `--patch-module`. now everything should work as expected we will get in `build/libs` jpf-nas.jar and jpf-nas-classes.jar

Mahmoud-Khawaja committed:
feat: added github actions workflow

cyrille-artho commented on .github/workflows/gradle.yml:
```diff
@@ -0,0 +1,38 @@
+          echo "extensions = $HOME/jpf-core" >> $HOME/.jpf/site.properties
+
+      - name: Build and test jpf-nas
+        run: ./gradlew clean build
```
The test action should use `./gradlew clean test`, as otherwise, only the build is run but not the tests.

cyrille-artho commented on buildSrc/build.gradle:
```diff
@@ -0,0 +1,8 @@
+repositories {
```
Is it necessary to have a second source file tree?

cyrille-artho reviewed (changes requested):
Let's discuss the `buildSrc` part, as I don't understand why it is needed. `Jar.java` seems to have been part of an earlier version of `jpf-nas`, but it is not in the current version for some reason.

Mahmoud-Khawaja committed:
Update gradle.yml

cyrille-artho commented on .gitignore:
```diff
@@ -0,0 +1,35 @@
+.idea/
+*.iml
+*.ipr
+*.iws
```
Add *.xml or workspace.xml

cyrille-artho reviewed (changes requested):
Remove workspace.xml

Mahmoud-Khawaja committed:
Update .gitignore

Mahmoud-Khawaja committed:
Delete .idea/workspace.xml

cyrille-artho commented:
Thanks for the update. It seems the two failing tests still fail. The tests will not be run as a GitHub workflow until after we have accepted a PR, so please let me know when the tests pass on your computer.

Mahmoud-Khawaja committed:
add compilation for examples

cyrille-artho commented:
I can reproduce your results (3 passing, 2 failing tests).
As the next step, please check the structure of the generated files in `build` and the JAR files. Initially, we want to mirror the structure of the files from the Java 8 version. Later on, we can investigate if some adjustments are needed (due to the Java module system, which may require us to change the location of some files).

Mahmoud-Khawaja committed:
fix for gralde to include the peers and jar exampes, main and test

Mahmoud-Khawaja committed:
fix for class path

Mahmoud-Khawaja committed:
modified ServerSocket.java

Mahmoud-Khawaja committed:
fix for the modelJar

cyrille-artho commented:
Hi,
Thank you for working on this. Most of the build now works (as most files are compiled and seem to be in the right place), but I also still get five failing tests. The files in `src/examples` are not built when I try to run the build. However, oddly in the Java 8 build they are not included in the JAR files; it is enough to have the bytecode in `build/examples` to use them for the unit tests.

Overall, I think there are two things to add/fix:

1. JAR dependency in the build

One problem is that that JAR files are not built as per of `./gradlew test`. Without the JAR files in `build/libs`, I get the following errors for the tests:
```
[WARNING] unknown classpath element: /Users/cartho/jpf/jpf-nas-mahmoud/build/model-classes
[WARNING] unknown classpath element: /Users/cartho/jpf/jpf-nas-mahmoud/build/libs/jpf-nas-classes.jar
[WARNING] unknown classpath element: /Users/cartho/jpf/jpf-nas-mahmoud/build/classes/java/examples
```
Manually running `./gradlew jar` before `./gradlew test` reduces these errors to one:
```
[WARNING] unknown classpath element: /Users/cartho/jpf/jpf-nas-mahmoud/build/classes/java/examples
```
The content of the JAR files seems (in terms of files contained therein) mostly identical. The two versions of `jpf-nas.jar` contain the same files both for Java 8 and Java 11. The Java 11 version of `jpf-nas-classes.jar` also has the same files except for `java/net/ServerSocket$1.class` that is missing, which is likely due to a feature in Java bytecode for Java 11 that requires fewer synthetic inner classes in the compiled code.

2. Missing compilation of examples

You have added a build rule for the examples, but for some reason, it is not triggered. Please check. Under Java 8, we have the examples compiled to `build/examples` without being included in the JAR files. Including them in the JAR files is fine, so you can keep that part of you like.

# Summary

Please check the build dependencies so the JAR files are build as a dependency of "test", and check how to compile the files in `examples`.

Mahmoud-Khawaja committed:
created PlainSocketImple peer class

Mahmoud-Khawaja committed:
fix build

Mahmoud-Khawaja committed:
Implemented comprehensive Java 11 support for JPF-NAS socket simulation:
  - JPF_java_net_PlainSocketImpl: Intercepts socket initialization, prevents ExtendedSocketOptions chain
  - JPF_java_net_InetAddress: Handles hostname resolution and address management
  - JPF_java_net_SocketCleanable: Bypasses Java 11 automatic resource cleanup
  - JPF_java_lang_ref_Reference: Supports reachabilityFence() for reference management
  - JPF_jdk_internal_ref_PhantomCleanable: Handles phantom reference cleanup
  - JPF_jdk_internal_ref_CleanerFactory: Manages cleaner system integration

Mahmoud-Khawaja committed:
removed unused functions in peer classes and reduced warnings. and fix for testTimedoutRead

Mahmoud-Khawaja committed:
use direct implementation

Mahmoud-Khawaja committed:
refactored the file structure to align with jpf-core

Mahmoud-Khawaja committed:
fix build

ahmed-n-abdeltwab commented:
Hi @Mahmoud-Khawaja and @cyrille-artho 

I’ve been exploring the project over the past few days, and I wanted to share a quick update on what I’ve done so far. I started by cloning and building the `jpf-core` repository successfully. When I moved on to `jpf-nas`, I ran into an issue where the build expected a `.jpf` directory in my home folder with a `site.properties` file. Since I preferred not to create that directory directly in my home for some reasons, I modified the code slightly to read the JPF core path from an environment variable instead. That allowed me to build and test the project using this command:

```bash
./gradlew clean test -DJPF_CORE_PATH=$HOME/IdeaProjects/jpf-core
```

The build ran, and most of the tests passed, but two of them failed: `testTimedoutAccept` and `testDeadlockForBlockedFinalizer_MulitProcessVM`. I plan to look further into why these specific tests are failing. So far, everything else is working as expected, and this process has helped me get a better feel for the project structure and build system.

If there’s anything you’d recommend I focus on next or any feedback on the approach I took, I’d really appreciate it. Thanks again for all your help and support!

cyrille-artho commented:
Hi Ahmed,
Thank you for looking into this.
If I remember correctly, JPF has a fallback mechanism to look for `jpf-core` in a sibling directory of an extension. If you have jpf-core in `~/jpf/jpf-core` and jpf-nas in `~/jpf/jpf-nas`, then maybe the build works without setting an environment variable.
You can also try using a local `site.properties`, as JPF has an option `+jpf.site=...` that you can use to override the default (which is `~/.jpf/site.properties` and, as you have noted, not ideal for testing JPF itself).

ahmed-n-abdeltwab commented:
Thank you, @cyrille-artho, for the clarification! I really appreciate the guidance. I'll give that setup a try since, as you mentioned, it's not the best option for testing JPF itself. I'll keep you posted if I run into anything else.

Mahmoud-Khawaja committed:
fix for the unkown class path
